src/cuda/nn/nll_loss.cu

```diff
@@ -0,0 +1,124 @@
+
+#include "nn_utils.cu"
+
+template <typename T>
+__global__ void __nll_loss_1D_index_reduce_kernel(
+    const T* src,
+    len_t trg,
+    len_t n,
+    T* total_rdx,
+    int ignore_idx,
+    reduxtype reduxop
+) {
+  if (static_cast<int>(trg) == ignore_idx) {
+    *total_rdx = T(0);
+    return;
+  }
+  // NONE means we didn't reduce at all
+  // so it's not actually an option here
+  if (reduxop == RDX_SUM) {
+      *total_rdx = -src[trg];
+  } else {
+      *total_rdx = -src[trg] / T(n);
+  }
+}
+
+template<typename T>
+void __nll_loss_1D_index_reduce(
+  void* cudnn_handle,
+  void* src,
+  len_t trg,
+  len_t n,
+  void* rdx,
+  int ignore_idx,
+  reduxtype reduxop
+) {
+  cudaStream_t stream;
+  CUDNN_ASSERT(cudnnGetStream(static_cast<cudnnHandle_t>(cudnn_handle), &stream));
+  const auto _src = static_cast<T*>(src);
+  const auto _rdx = static_cast<T*>(rdx);
+  __nll_loss_1D_index_reduce_kernel<T><<<1,1,0,stream>>>(_src, trg, n, _rdx, ignore_idx, reduxop);
+  return;
+}
+
+extern "C" void nll_loss_1D_index_reduce(
```

Comment:
I need to add a `DtoH` memcpy here and we can just return the float directly. I think synchronization is fine here because if you're printing loss then you need the value in hand.